### PR TITLE
Limit torch_agent.Batch to tensors

### PR DESCRIPTION
**Pull request by stephenroller**
**Patch description**
Where possible, limit the attributes of our Batch object to only tensors. Additionally, delay cudafying until late and switch to using batch.batchsize for many operations.

This is in preparation of a fresh attempt of Background Preprocessing.

**Testing steps**
CI

### CODE REVIEW

File: docs/source/tutorial_metrics.md [hunk 1]
```diff
@@ -421,11 +421,11 @@ please [file an issue on GitHub](https://github.com/facebookresearch/ParlAI/issu
 | ----------------------- | ------------ |
 | `accuracy`              | Exact match text accuracy |
 | `bleu-4`                | BLEU-4 of the generation, under a standardized (model-independent) tokenizer |
+| `clen`                  | Average length of context in number of tokens |
 | `clip`                  | Fraction of batches with clipped gradients |
-|`context_truncate`       | Ratio of samples' contexts being truncated per batch |
-|`context_length`         | Average length of context tokens per batch |
 | `ctpb`                  | Context tokens per batch |
 | `ctps`                  | Context tokens per second |
+| `ctrun`                 | Fraction of samples with some context truncation |
```
Comment (klshuster) [inline docs/source/tutorial_metrics.md:428]: nit: can we have it be `ctrunc`; this reads to me like `ct run` which is a bit confusing

File: docs/source/tutorial_metrics.md [hunk 2]
```diff
@@ -435,12 +435,12 @@ please [file an issue on GitHub](https://github.com/facebookresearch/ParlAI/issu
...
 | `intradistinct-1`, `intradictinct-2` | Fraction of n-grams unique _within_ each utterance |
 | `jga`                   | Joint Goal Accuracy |
-|`label_length`           | Average length of label tokens per batch |
-|`label_truncate`         | Ratio of samples' labels being truncated per batch |
+| `llen`                  | Average length of label in number of tokens |
 | `loss`                  | Loss |
 | `lr`                    | The most recent learning rate applied |
 | `ltpb`                  | Label tokens per batch |
 | `ltps`                  | Label tokens per second |
+| `ltrun`                 | Fraction of samples with some label truncation |
```
Comment (klshuster) [inline docs/source/tutorial_metrics.md:443]: ditto. totally fine to leave it if we want to keep abbreviations 5 chars or less

File: parlai/agents/hred/hred.py [hunk 3]
```diff
@@ -129,26 +129,15 @@ def batchify(self, obs_batch, sort=True):
...
-                batch_context_vec.append(torch.tensor(context_vec, device=self.device))
-
-        padded_context_vec = torch.nn.utils.rnn.pad_sequence(
-            batch_context_vec, batch_first=True
-        ).squeeze(1)
-        return (
-            padded_context_vec,
-            torch.tensor(hist_lens, dtype=torch.long, device=self.device),
+        # sum here is list concat, not addition
+        context_vec, hist_lens_ = self._pad_tensor(
```
Comment (klshuster) [inline parlai/agents/hred/hred.py:133]: perhaps unrelated, but padded tensor returning the lengths as simply the lengths of the input lists is not super intuitive, especially when an optional argument is the pad token; ideally you'd want to return the lengths of the unpadded input lists

is this why you recompute `hist_lens` below?

File: parlai/agents/test_agents/test_agents.py [hunk 4]
```diff
@@ -114,18 +114,16 @@ def train_step(self, batch):
         """
         Return confirmation of training.
         """
-        return Output(['Training {}!'.format(i) for i in range(len(batch.text_vec))])
+        return Output(['Training {}!'.format(i) for i in range(batch.batchsize)])
```
Comment (klshuster) [inline parlai/agents/test_agents/test_agents.py:117]: nit: can use f string here

File: parlai/agents/test_agents/test_agents.py [hunk 5]
```diff
@@ -114,18 +114,16 @@ def train_step(self, batch):
...
         """
         Return confirmation of evaluation.
         """
         return Output(
             [
-                'Evaluating {} (responding to {})!'.format(
-                    i, batch.observations[i]['text']
-                )
-                for i in range(len(batch.text_vec))
+                'Evaluating {} (responding to {})!'.format(i, batch.text_vec.tolist())
```
Comment (klshuster) [inline parlai/agents/test_agents/test_agents.py:125]: ditto

File: parlai/core/torch_agent.py [hunk 6]
```diff
@@ -123,8 +113,12 @@ def __init__(
         valid_indices=None,
         candidates=None,
         candidate_vecs=None,
+        reward=None,
```
Comment (klshuster) [inline parlai/core/torch_agent.py:116]: where is this `reward` coming from?

File: parlai/core/torch_agent.py [hunk 7]
```diff
@@ -137,10 +131,50 @@ def __init__(
...
+            _label_original_length=_label_original_length,
+            _label_truncate_rate=_label_truncate_rate,
             **kwargs,
         )
 
+    def to(self, dev):
+        """
+        Move all tensors in the batch to a device.
+
+        Happens in place. Note that valid_indices and fields starting with an
```
Comment (klshuster) [inline parlai/core/torch_agent.py:145]: i like these semantics. can we make that more clear in the Batch object description? specifically underscored fields being exempt

File: parlai/core/torch_agent.py [hunk 8]
```diff
@@ -96,22 +88,20 @@ class Batch(AttrDict):
...
     candidates: Optional[List[List[str]]]
     candidate_vecs: Optional[List[List[torch.LongTensor]]]
     image: Optional[List[Any]]
-    observations: Optional[List[Message]]
+    _context_original_length: Optional[torch.LongTensor]
+    _context_truncate_rate: Optional[torch.LongTensor]
+    _label_original_length: Optional[torch.LongTensor]
+    _label_truncate_rate: Optional[torch.LongTensor]
 
     def __init__(
```
Comment (klshuster) [inline parlai/core/torch_agent.py:106]: perhaps we can warn here if we're passing in a non-tensor?

File: parlai/utils/torch.py [hunk 9]
```diff
@@ -65,11 +65,9 @@ def atomic_save(state_dict: Any, path: str) -> None:
 def padded_tensor(
     items: List[Union[List[int], torch.LongTensor]],
     pad_idx: int = 0,
-    use_cuda: bool = False,
```
Comment (klshuster) [inline parlai/utils/torch.py:68]: this will almost certainly break some internal stuff but i suppose those can be dealt with individually

**Review by klshuster — APPROVED** [review]
sweet

File: parlai/core/torch_agent.py [hunk 10]
```diff
@@ -137,10 +131,50 @@ def __init__(
...
+        output = ['Batch({']
+        for key in sorted(self.keys()):
+            value = self[key]
+            if key == 'observations' and value is None:
+                output.append(f'  {key}: {value} (use --debug to include),')
+            elif value is None:
+                output.append(f'  {key}: {value},')
+            elif isinstance(value, torch.Tensor):
+                typ = value.type().replace("torch.", "")
+                shape = ", ".join(str(s) for s in value.shape)
```
Comment (EricMichaelSmith) [inline parlai/core/torch_agent.py:170]: ❤️ ❤️

File: projects/style_gen/classifier.py [hunk 11]
```diff
@@ -166,15 +166,6 @@ def _format_interactive_output(self, probs, prediction_id):
             )
         return preds
 
-    def get_labels_field(self, observations):
```
Comment (EricMichaelSmith) [inline projects/style_gen/classifier.py:169]: Yay for simplification!

File: parlai/agents/hred/hred.py (hunk 3 above)
Comment (stephenroller) [reply parlai/agents/hred/hred.py:133]: idk bro I'm just trying to get that one to work lol

File: parlai/core/torch_agent.py (hunk 8 above)
Comment (stephenroller) [reply parlai/core/torch_agent.py:106]: For specifically these 4 or others? We have non-tensors (bools, Nones, etc).

File: parlai/core/torch_agent.py (hunk 6 above)
Comment (stephenroller) [reply parlai/core/torch_agent.py:116]: Some teachers provide it now and we batchify it. It's used by Unlikelihood and others.

File: parlai/agents/hred/hred.py (hunk 3 above)
Comment (klshuster) [reply parlai/agents/hred/hred.py:133]: my b i didn't realize this was in the `hred` agent

File: parlai/core/torch_agent.py (hunk 6 above)
Comment (klshuster) [reply parlai/core/torch_agent.py:116]: could you include it in the `Batch` docstring?

File: parlai/core/torch_agent.py (hunk 8 above)
Comment (klshuster) [reply parlai/core/torch_agent.py:106]: wait so... we're not fully getting rid of non-tensors here?

Comment (stephenroller) [reply parlai/core/torch_agent.py:106]: Nope. I benchmarked them and they're not painful. Only the complex objects are.

Comment (klshuster) [reply parlai/core/torch_agent.py:106]: how complex is complex? does this mean I can just make my own batch object and keep strings in there?

Comment (stephenroller) [reply parlai/core/torch_agent.py:106]: Ya I mean, we don't have any hard limitation. You can even do full observations if you want, you'll just pay a penalty with background workers. But any Batch object you _manually_ want to add things to is still allowed.

I'll leave notes in batch descriptions
